pkg/fqdn/cache.go

```diff
@@ -0,0 +1,191 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fqdn
+
+import (
+	"bytes"
+	"net"
+	"sort"
+	"time"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var DefaultDNSCache = NewDNSCache()
+
+// cacheEntry objects are immutable once created.
+// They hold an the DNS data passed in with DNSCache.Update, nominally equating
+// to a DNS lookup. cacheEntry objects are managed by DNSCache and are not
+// expected to be returned.
+type cacheEntry struct {
+	Name           string
+	LookupTime     time.Time
+	ExpirationTime time.Time
+	TTL            int
+	IPs            []net.IP
+}
+
+func (entry *cacheEntry) isExpiredBy(pointInTime time.Time) bool {
+	return pointInTime.After(entry.ExpirationTime)
+}
+
+// ipEntries maps a unique IP to the cacheEntry that provides it in .IPs.
+// Multiple IPs may point to the same cacheEntry, or they may all be different.
+// Crucially, an IP may be present in a cacheEntry but the IP in ipEntries
+// points to another cacheEntry. This is because the second cacheEntry has a
+// later expiration for this specific IP, and may not include the other IPs
+// provided by the first entry.
+// The DNS name in the entries is not checked, but is assumed to be the same
+// for all entries.
+// Note: They are guarded by the DNSCache mutex.
+type ipEntries map[string]*cacheEntry
+
+// getIPs returns a sorted list of non-expired unique IPs.
+// This needs a read-lock
+func (s ipEntries) getIPs(now time.Time) []net.IP {
+	ips := make([]net.IP, 0, len(s)) // worst case size
+	for ip, entry := range s {
+		if entry != nil && !entry.isExpiredBy(now) {
+			ips = append(ips, net.ParseIP(ip))
+		}
+	}
+
+	return keepUniqueIPs(ips) // sorts IPs
+}
+
+// DNSCache manages DNS data that will expire after a certain TTL. Information
+// is tracked per-IP address, retaining the latest-expiring DNS data for each
+// address.
+// For most real-world DNS data, the entry per name remains small because newer
+// lookups replace older ones. Large TTLs may cause entries to grow if many
+// unique IPs are returned in separate lookups.
+// Redundant or expired entries are removed on insert.
+// Lookups check for expired entries.
+type DNSCache struct {
+	lock.RWMutex
+
+	// forward DNS lookups name -> IPEntries
+	// IPEntries maps IP -> entry that provides it. An entry may provide multiple IPs.
+	forward map[string]ipEntries
+}
+
+// NewDNSCache returns an initialized DNSCache
+func NewDNSCache() *DNSCache {
+	c := &DNSCache{
+		forward: make(map[string]ipEntries),
+	}
+
+	return c
+}
+
+// Update inserts a new entry into the cache.
+// After insertion cache entries for name are expired and redundant entries
+// evicted. This is O(number of new IPs) for eviction, and O(number of IPs for
+// name) for expiration.
+func (c *DNSCache) Update(lookupTime time.Time, name string, ips []net.IP, ttl int) {
```

Comment:
I'm wary about providing a time to this function, which then acquires the lock - this could mean that we call Update, the lock is held for a long time for the DNSCache by some other goroutine, and then we enter this function, which can result in the TTL already being hit for objects we are updating. Not 100 % sure how valid this concern is. Why not make the `lookupTime` the `time.Now()` from right after the lock is acquired on line 106?

Reply:
I bounced around on this one, especially because `Lookup` doesn't take a time. I settled on this for a simple reason: The data source may update the cache after a non-trivial delay. This would then impact the calculated `ExpirationTime` and the validity of the data.
If you look at the DNSPoller class, in `LookupUpdateDNS`, you can see this. It naively does all DNS lookups and updates the cache after all are done*. In that case, the outcome is that some DNS queries are expired sooner than they should be. If we used time.Now() in Update, some would expire later than they should. Some variant on this is likely when ingesting DNS data from other sources so I effectively deferred this to the caller.

* This is quite bad. I'll need to improve on it in a later PR. A single bad DNS lookup may delay updates quite a bit. Unfortunately, the poller has two constraints here: don't spam the DNS servers, don't spam new generated policy (and regenerations). The middle part, however, can be cleaner.

Reply:
Fair enough - agreed that this can be improved, but I am all about iterative improvements over time, so keeping as-is makes sense for now.